packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart

```diff
@@ -24,8 +28,138 @@ class DeepLinksScreen extends Screen {
 
   @override
   Widget build(BuildContext context) {
-    return const Center(
-      child: Text('TODO: build deep link validation tool'),
+    return const DeepLinkPage();
+  }
+}
+
+class DeepLinkPage extends StatefulWidget {
+  const DeepLinkPage({super.key});
+
+  @override
+  State<DeepLinkPage> createState() => _DeepLinkPageState();
+}
+
+class _DeepLinkPageState extends State<DeepLinkPage>
+    with
+        AutoDisposeMixin,
+        SingleTickerProviderStateMixin,
+        ProvidedControllerMixin<DeepLinksController, DeepLinkPage> {
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    if (!initController()) return;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return ListView(
+      children: [
+        Row(
+          children: [
+            Expanded(
+              child: Text(
+                'All deep links',
+                style: Theme.of(context).textTheme.bodyLarge,
+              ),
+            ),
+            InkWell(
+              onTap: () {
+                controller.bundleByDomain = !controller.bundleByDomain;
+              },
+              child: const Text('Bundle by domain'),
+            ),
+            const SizedBox(width: denseSpacing),
+            SearchBar(
+              leading: const Icon(Icons.search),
+              onChanged: (value) {
+                controller.searchContent = value;
+              },
+              constraints: BoxConstraints.tight(
+                Size(defaultSearchFieldWidth, defaultTextFieldHeight),
+              ),
+            ),
+          ],
+        ),
+        const SizedBox(height: denseSpacing),
+        buildDataTable(context),
+      ],
+    );
+  }
+
+  Widget buildDataTable(BuildContext context) {
+    return ValueListenableBuilder<List<LinkData>>(
+      valueListenable: controller.linkDatasNotifier,
+      builder: (context, linkDatas, _) {
+        final bool bundleByDomain = controller.bundleByDomain;
+
+        final List<DataRow> rows = [
+          for (var i = 0; i < linkDatas.length; i++)
+            buildRow(
+              context,
+              linkDatas[i],
+              color: MaterialStateProperty.all<Color>(
+                alternatingColorForIndex(i, Theme.of(context).colorScheme),
+              ),
+            ),
+        ];
+
+        return DataTable(
+          columns: const [
+            DataColumn(label: Text('OS')),
+            DataColumn(label: Text('Scheme')),
+            DataColumn(label: Text('Domain')),
+            DataColumn(label: Text('Path')),
+          ],
+          dataRowMinHeight: bundleByDomain ? bundledDataRowHeight : null,
+          dataRowMaxHeight: bundleByDomain ? bundledDataRowHeight : null,
+          rows: rows,
+        );
+      },
     );
   }
 }
+
+DataRow buildRow(
```

Comment:
same here. Make this a helper widget instead of having a method that returns a widget

Reply:
the `DataRow` is not a widget, it's a data structure used in `DataTable`